R/conditional-transform.R

```diff
@@ -0,0 +1,101 @@
+#' @param default_value a default value to use if none of the conditions are true (default: `NA`)
+#' @param type a character that is either "categorical", "text", "numeric" what type of output should be returned?
+#' @param categories if `type="categorical"`, these are all of the categories that should be in the resulting variable
+#'
+#' @return either a factor (if `type="categorical"`) or character (if `type="text"`)
+#' @examples
+#' \dontrun{
+#' ds$cat_opinion <- conditionalTransform(ds$pet1 == 'Cat' ~ ds$Opinion1,
+#'                                        ds$pet2 == 'Cat' ~ ds$Opinion2,
+#'                                        ds$pet3 == 'Cat' ~ ds$Opinion3)
+#' ds$dot_opinion <- conditionalTransform(ds$pet1 == 'Dog' ~ ds$Opinion1,
+#'                                        ds$pet2 == 'Dog' ~ ds$Opinion2,
+#'                                        ds$pet3 == 'Dog' ~ ds$Opinion3)
+#' }
+#'
+#' @export
+conditionalTransform <- function (..., default_value=NA, type="categorical", categories=NULL) {
+    formulas <- list(...)
+    n <- length(formulas)
+
+    if (n == 0) {
+        halt("no conditions have been supplied")
+    }
+    if (!type %in% c("categorical", "text", "numeric")){
+        halt("type must be either ", dQuote("categorical"), dQuote("text"), " or ", dQuote("numeric"))
+    }
+    if (type != "categorical" & !is.null(categories)){
+        halt("type is not ", dQuote("categorical"), " ignoring ", dQuote("categories"))
+    }
+
+    cases <- vector("list", n)
+    values <- vector("list", n)
+    for (i in seq_len(n)) {
+        formula <- formulas[[i]]
+        if (!inherits(formula, "formula") || length(formula) != 3) {
+            halt("The case provided is not a formula: ", deparseAndFlatten(substitute(list(...))[[i + 1]]))
+        }
+
+        # TODO: switch ot the more lightweight rlang::bare_eval()?
+        cases[[i]] <- eval(formula[[2]])
+        if (!inherits(cases[[i]], "CrunchLogicalExpr")) {
+            halt("The LHS provided is not a CrunchLogicalExpr: ", deparseAndFlatten(substitute(list(...))[[i + 1]][[2]]))
+        }
+        values[[i]] <- eval(formula[[3]])
+    }
+
+    n_rows <- nrow(CrunchDataset(crGET(datasetReference(cases[[1]]))))
+
+    case_indices <- lapply(cases, which)
+
+    # dedpulicate indices, favoring the first observation
+    case_indices <- lapply(seq_along(case_indices), function(i) {
+        setdiff(case_indices[[i]], unlist(case_indices[seq_len(i-1)]))
+    })
+
+    values_to_fill <- Map(function(ind, var) {
+        if (inherits(var, "CrunchVariable")) {
+            # grab the variable contents at inds
+            return(as.vector(var[ind]))
+        } else {
+            # if var isn't a crunchvariable, just return var
+            return(var)
+        }
+    }, ind = case_indices, var = values)
+
+    # setup NAs for as default
+    result <- rep(default_value, n_rows)
+
+    for (i in seq_along(case_indices)) {
+        if (type == "numeric") trans <- as.numeric
```

Comment:
(1) the issue I ran into is when we were trying to store elements from factors, I needed to `as.character` them in order to get the level instead of the level id:
```
> foo <- factor(c("A", "B"))
> bar <- c("c","d")
> baz <- rep(NA, 2)
> baz[1] <- foo[1]
> baz[2] <- bar[2]
> baz
[1] "1" "d"
> baz <- rep(NA, 2)
> baz[1] <- as.character(foo[1])
> baz[2] <- bar[2]
> baz
[1] "A" "d"
```
We could just limit this transformation to categoricals that get used, and use coercion otherwise. Although if we keep the type specifications we should probably at least check the types at the end if not converting them specifically.

(2) done